### PR TITLE
Add OKR metrics to Desktop and Mobile products

### PR DESCRIPTION
**Pull request by bochocki**
Adds OKR definitions (Engagement Rate, Retention Rate, New Profile Retention Rate) to Metric Hub ([DS-3762](https://mozilla-hub.atlassian.net/browse/DS-3762)).

### CODE REVIEW

File: definitions/firefox_desktop.toml [hunk 1]
```diff
@@ -2051,6 +2089,19 @@ description = "Retention Outcomes of Desktop New Profiles Over their First 112 D
 friendly_name = "Desktop Cohorts Daily Retention"
 client_id_column = "NULL"
 
+[data_sources.desktop_engagement_view]
+from_expression = "moz-fx-data-shared-prod.telemetry.desktop_engagement"
```
Comment (bani) [inline definitions/firefox_desktop.toml:2093]: Should we add the is_desktop filter to this one for consistency with mobile?

File: definitions/fenix.toml [hunk 2]
```diff
@@ -642,3 +680,25 @@ from_expression = """(
...
+client_id_column = "NULL"
+
+[data_sources.fenix_retention_view]
+from_expression = """(
+    SELECT *
+     FROM `moz-fx-data-shared-prod.fenix.retention`
+    WHERE is_mobile
+)"""
+submission_date_column = "metric_date"
+description = "4th Week Retention Outcomes of Fenix New and Existing Users."
```
Comment (irrationalagent) [inline definitions/fenix.toml:702]: nit, 


```suggestion
description = "Aggregated 4th Week Retention Outcomes of Fenix New and Existing Users."
```

File: definitions/firefox_desktop.toml [hunk 3]
```diff
@@ -2051,6 +2089,19 @@ description = "Retention Outcomes of Desktop New Profiles Over their First 112 D
...
+submission_date_column = "submission_date"
+description = "Aggregated DAU, WAU, and MAU by different attributes for engagement ratio calculation."
+friendly_name = "Desktop Engagement"
+client_id_column = "NULL"
+
+[data_sources.desktop_retention_view]
+from_expression = "moz-fx-data-shared-prod.telemetry.desktop_retention"
+submission_date_column = "metric_date"
+description = "4th Week Retention Outcomes of Desktop New and Existing Users."
```
Comment (irrationalagent) [inline definitions/firefox_desktop.toml:2102]:
```suggestion
description = "Aggregated 4th Week Retention Outcomes of Desktop New and Existing Users."
```

File: definitions/firefox_ios.toml [hunk 4]
```diff
@@ -530,3 +567,25 @@ friendly_name = "iOS Funnel Retention Metrics"
...
+client_id_column = "NULL"
+
+[data_sources.firefox_ios_retention_view]
+from_expression = """(
+    SELECT *
+     FROM `moz-fx-data-shared-prod.firefox_ios.retention`
+    WHERE is_mobile
+)"""
+submission_date_column = "metric_date"
+description = "4th Week Retention Outcomes of Firefox iOS New and Existing Users."
```
Comment (irrationalagent) [inline definitions/firefox_ios.toml:589]:
```suggestion
description = "Aggregated 4th Week Retention Outcomes of Firefox iOS New and Existing Users."
```

File: definitions/focus_android.toml [hunk 5]
```diff
@@ -314,3 +351,24 @@ experiments_column_type = "simple"
...
+client_id_column = "NULL"
+
+[data_sources.focus_android_retention_view]
+from_expression = """(
+    SELECT *
+     FROM `moz-fx-data-shared-prod.focus_android.retention`
+    WHERE is_mobile
+)"""
+submission_date_column = "metric_date"
+description = "4th Week Retention Outcomes of Focus Android New and Existing Users."
```
Comment (irrationalagent) [inline definitions/focus_android.toml:372]:
```suggestion
description = "Aggregated 4th Week Retention Outcomes of Focus Android New and Existing Users."
```

File: definitions/focus_ios.toml [hunk 6]
```diff
@@ -321,3 +357,25 @@ from_expression = """(
...
+client_id_column = "NULL"
+
+[data_sources.focus_ios_retention_view]
+from_expression = """(
+    SELECT *
+     FROM `moz-fx-data-shared-prod.focus_ios.retention`
+    WHERE is_mobile
+)"""
+submission_date_column = "metric_date"
+description = "4th Week Retention Outcomes of Focus iOS New and Existing Users."
```
Comment (irrationalagent) [inline definitions/focus_ios.toml:379]:
```suggestion
description = "Aggregated 4th Week Retention Outcomes of Focus iOS New and Existing Users."
```

File: definitions/klar_android.toml [hunk 7]
```diff
@@ -192,3 +229,25 @@ from_expression = "mozdata.search.mobile_search_clients_engines_sources_daily"
...
+client_id_column = "NULL"
+
+[data_sources.klar_android_retention_view]
+from_expression = """(
+    SELECT *
+     FROM `moz-fx-data-shared-prod.klar_android.retention`
+    WHERE app_name = "Klar Android"
+)"""
+submission_date_column = "metric_date"
+description = "4th Week Retention Outcomes of Klar Android New and Existing Users."
```
Comment (irrationalagent) [inline definitions/klar_android.toml:251]:
```suggestion
description = "Aggregated 4th Week Retention Outcomes of Klar Android New and Existing Users."
```

File: definitions/klar_ios.toml [hunk 8]
```diff
@@ -196,3 +233,25 @@ from_expression = """(
...
+client_id_column = "NULL"
+
+[data_sources.klar_ios_retention_view]
+from_expression = """(
+    SELECT *
+     FROM `moz-fx-data-shared-prod.klar_ios.retention`
+    WHERE app_name = "Klar iOS"
+)"""
+submission_date_column = "metric_date"
+description = "4th Week Retention Outcomes of Klar iOS New and Existing Users."
```
Comment (irrationalagent) [inline definitions/klar_ios.toml:255]:
```suggestion
description = "Aggregated 4th Week Retention Outcomes of Klar iOS New and Existing Users."
```

File: definitions/multi_product.toml [hunk 9]
```diff
@@ -133,3 +171,25 @@ submission_date_column = "submission_month"
...
+client_id_column = "NULL"
+
+[data_sources.mobile_retention_view]
+from_expression = """(
+    SELECT *
+     FROM `moz-fx-data-shared-prod.telemetry.mobile_retention`
+    WHERE is_mobile
+)"""
+submission_date_column = "metric_date"
+description = "4th Week Retention Outcomes of Mobile New and Existing Users."
```
Comment (irrationalagent) [inline definitions/multi_product.toml:193]:
```suggestion
description = "Aggregated 4th Week Retention Outcomes of Mobile New and Existing Users."
```

**Review by irrationalagent — APPROVED** [review]

File: definitions/firefox_desktop.toml (hunk 1 above)
Comment (bochocki) [reply definitions/firefox_desktop.toml]: Good call! I made those changes.

**Review by bani — APPROVED** [review]

**Comment by bochocki** [comment]
I don't know if there's a better way to do this, but I wrote a short Python script to try to validate that these metrics are working as expected after merging. The script is below.

All metrics could be queried successfully using this approach.

```shell
!pip install mozanalysis -qq

```

```python3
import pandas as pd

from google.cloud import bigquery
from google.colab import auth
from mozanalysis.config import ConfigLoader
from textwrap import dedent

auth.authenticate_user()

def fetch_metric_hub_definition(start_date, end_date, metric_slug, app_name, bq_project):
    start_date = pd.to_datetime(start_date).date()
    end_date = pd.to_datetime(end_date).date()

    # Set useful attributes based on the Metric Hub definition
    metric = ConfigLoader.get_metric(metric_slug=metric_slug, app_name=app_name)
    submission_date_column = metric.data_source.submission_date_column

    # Modify the metric source table string so that it formats nicely in the query.
    from_expression = metric.data_source._from_expr.replace("\n", "\n" + " " * 15)

    query = dedent(
        f"""
        SELECT {submission_date_column},
               {metric.select_expr} AS value
          FROM {from_expression}
         WHERE {submission_date_column} BETWEEN '{start_date}' AND '{end_date}'
         GROUP BY {submission_date_column}
        """
    )

    print(
        "\n",
        "\n" + "-" * 110,
        "\n" + f" Querying for '{app_name}.{metric_slug}' ".center(110, "-"),
        f"\n{query}"
    )
    df = bigquery.Client(project=bq_project).query(query).to_dataframe()

    # ensure submission_date has type 'date'
    df[submission_date_column] = pd.to_datetime(df[submission_date_column]).dt.date

    return df
```

```python3
for app_name in ["firefox_desktop", "multi_product", "fenix", "firefox_ios", "focus_android", "focus_ios", "klar_android", "klar_ios"]:
    for metric_slug in ["engagement_rate_v1", "retention_rate_v1", "new_profile_retention_rate_v1"]:
        df = fetch_metric_hub_definition(
            start_date="2024-03-14",
            end_date="2024-03-21",
            metric_slug=metric_slug,
            app_name=app_name,
            bq_project="moz-fx-data-bq-data-science"
        )
        display(df.set_index(df.columns[0]).transpose())
```

**Pull request merged**